File: src/javascript/lib/list.js
import Kernel from './kernel';
import * as Core from './core';

let List = {};

List.delete = function(list, item){
  let new_value = [];
  let value_found = false;

  for(let x of list){
    if(x === item && value_found !== false){
      new_value.push(x);
      value_found = true;
    }else if(x !== item){
      new_value.push(x);
    }
  }

  return Core.SpecialForms.list(...new_value);
};

List.delete_at = function(list, index){
  let new_value = [];

  for(let i = 0; i < list.length; i++){
    if(i !== index){
      new_value.push(list[i]);
    }
  }

  return Core.SpecialForms.list(...new_value);
};

List.duplicate = function(elem, n){
  let new_value = [];

  for (var i = 0; i < n; i++) {
    new_value.push(elem);
  }

  return Core.SpecialForms.list(...new_value);
};

List.first = function(list){
  return list[0];
};

List.flatten = function(list, tail = Core.SpecialForms.list()){
  let new_value = [];

  for(let x of list){
    if(Kernel.is_list(x)){
      new_value = new_value.concat(List.flatten(x));
    }else{
      new_value.push(x);
    }
  }

  new_value = new_value.concat(tail);

  return Core.SpecialForms.list(...new_value);
};

List.foldl = function(list, acc, func){
  return list.reduce(func, acc);
};

List.foldr = function(list, acc, func){
  let new_acc = acc;

  for (var i = list.length - 1; i >= 0; i--) {
    new_acc = func(list[i], new_acc);
  }

  return new_acc;
};

List.insert_at = function(list, index, value){
  let new_value = [];

  for(let i = 0; i < list.length; i++){
    if(i === index){
      new_value.push(value);
      new_value.push(list[i]);
    }else{
      new_value.push(list[i]);
    }
  }

  return Core.SpecialForms.list(...new_value);
};

List.keydelete = function(list, key, position){
  let new_list = [];

  for(let i = 0; i < list.length; i++){
    if(!Kernel.match__qmark__(list[i][position], key)){
      new_list.push(list[i]);
    }
  }

  return Core.SpecialForms.list(...new_list);
};

List.keyfind = function(list, key, position, _default = null){

  for(let i = 0; i < list.length; i++){
    if(Kernel.match__qmark__(list[i][position], key)){
      return list[i];
    }
  }

  return _default;
};

List.keymember__qmark__ = function(list, key, position){

  for(let i = 0; i < list.length; i++){
    if(Kernel.match__qmark__(list[i][position], key)){
      return true;
    }
  }

  return false;
};

List.keyreplace = function(list, key, position, new_tuple){
  let new_list = [];

  for(let i = 0; i < list.length; i++){
    if(!Kernel.match__qmark__(list[i][position], key)){
      new_list.push(list[i]);
    }else{
      new_list.push(new_tuple);
    }
  }

  return Core.SpecialForms.list(...new_list);
};


List.keysort = function(list, position){
  let new_list = list;

  new_list.sort(function(a, b){
    if(position === 0){
      if(a[position].value < b[position].value){
        return -1;
      }

      if(a[position].value > b[position].value){
        return 1;
      }

      return 0;
    }else{
      if(a[position] < b[position]){
        return -1;
      }

      if(a[position] > b[position]){
        return 1;
      }

      return 0;
    }

  });

  return Core.SpecialForms.list(...new_list);
};

List.keystore = function(list, key, position, new_tuple){
  let new_list = [];
  let replaced = false;

  for(let i = 0; i < list.length; i++){
    if(!Kernel.match__qmark__(list[i][position], key)){
      new_list.push(list[i]);
    }else{
      new_list.push(new_tuple);
      replaced = true;
    }
  }

  if(!replaced){
    new_list.push(new_tuple);
  }

  return Core.SpecialForms.list(...new_list);
};

List.last = function(list){
  return list[list.length - 1];
};

List.replace_at = function(list, index, value){
  let new_value = [];

  for(let i = 0; i < list.length; i++){
    if(i === index){
      new_value.push(value);
    }else{
      new_value.push(list[i]);
    }
  }

  return Core.SpecialForms.list(...new_value);
};

List.update_at = function(list, index, fun){
  let new_value = [];

  for(let i = 0; i < list.count(); i++){
    if(i === index){
      new_value.push(fun(list.get(i)));
    }else{
      new_value.push(list.get(i));
    }
  }

  return new_value;
};

List.wrap = function(list){
  if(Kernel.is_list(list)){
    return list;
  }else if(list == null){
    return Core.SpecialForms.list();
  }else{
    return Core.SpecialForms.list(list);
  }
};

List.zip = function(list_of_lists){
  if(list_of_lists.length === 0){
    return Core.SpecialForms.list();
  }

  let new_value = [];
  let smallest_length = list_of_lists[0];

  for(let x of list_of_lists){
    if(x.length < smallest_length){
      smallest_length = x.length;
    }
  }

  for(let i = 0; i < smallest_length; i++){
    let current_value = [];
    for(let j = 0; j < list_of_lists.length; j++){
      current_value.push(list_of_lists[j][i]);
    }

    new_value.push(new Core.Tuple(...current_value));
  }

  return Core.SpecialForms.list(...new_value);
};

List.to_tuple = function(list){
  return new Core.Tuple(...list);
};

List.append = function(list, value){
  return Core.SpecialForms.list(...list.concat([value]));
};

List.prepend = function(list, value){
  return Core.SpecialForms.list(...[value].concat(list));
};

List.concat = function(left, right){
  return left.concat(right);
};

export default List;

File: priv/std_lib/list.ex
defmodule ElixirScript.List do
  @moduledoc false
  require JS

  def duplicate(data, size) do
    do_duplicate(data, size, [])
  end

  defp do_duplicate(_, 0, list) do
    list
  end

  defp do_duplicate(data, size, list) do
    do_duplicate(data, size - 1, list ++ [data])
  end

  def to_tuple(list) do
    JS.new(Bootstrap.Core.Tuple, list)
  end

  def wrap(list) when is_list(list), do: list
  def wrap(nil), do: []
  def wrap(term), do: [term]

  def append(list, term) do
    concat(list, [term])
  end

  def prepend(list, term) do
    concat([term], list)
  end

  def concat(list_a, list_b) do
    list_a.concat(list_b)
  end

  def first(list) do
    list[0]
  end

  def last(list) do
    list[length(list) - 1]
  end

  def delete(list, item) do
    do_delete(list, item, 0, [])
  end

  defp do_delete(list, item, current_index, new_list) do
    if current_index == length(list) do
      new_list
    else
      updated = case list[current_index] do
        ^item ->
          new_list
        _ ->
          new_list ++ [list[current_index]]
      end

      do_delete(list, item, current_index + 1, updated)
    end
  end

  def delete_at(list, index) do
    do_delete_at(list, index, 0, [])
  end

  defp do_delete_at(list, index, current_index, new_list) do
    if current_index == length(list) do
      new_list
    else
      updated = case current_index == index do
        true ->
          new_list
        _ ->
          new_list ++ [list[current_index]]
      end

      do_delete_at(list, index, current_index + 1, updated)
    end
  end

  def insert_at(list, index, value) do
    do_insert_at(list, index, value, 0, [])
  end

  defp do_insert_at(list, index, value, current_index, new_list) do
    if current_index == length(list) do
      new_list
    else
      updated = case current_index == index do
        true ->
          new_list ++ [value, list[current_index]]
        _ ->
          new_list ++ [list[current_index]]
      end

      do_insert_at(list, index, value, current_index + 1, updated)
    end
  end

  def replace_at(list, index, value) do
    do_replace_at(list, index, value, 0, [])
  end

  defp do_replace_at(list, index, value, current_index, new_list) do
    if current_index == length(list) do
      new_list
    else
      updated = case current_index == index do
        true ->
          new_list ++ [value]
        _ ->
          new_list ++ [list[current_index]]
      end

      do_replace_at(list, index, value, current_index + 1, updated)
    end
  end


  def update_at(list, index, func) do
    do_update_at(list, index, func, 0, [])
  end

  defp do_update_at(list, index, func, current_index, new_list) do
    if current_index == length(list) do
      new_list
    else
      updated = case current_index == index do
        true ->
          new_list ++ [func.(list[current_index])]
        _ ->
          new_list ++ [list[current_index]]
      end

      do_update_at(list, index, func, current_index + 1, updated)
    end
  end


  def foldl(list, acc, func) do
    do_foldl(list, acc, func, [])
  end

  def foldr(list, acc, func) do
    do_foldl(list.concat([]).reverse(), acc, func, [])
  end

  defp do_foldl([], acc, _, new_list) do
    { acc, new_list }
  end

  defp do_foldl(list, acc, func, new_list) do
    { acc, value } = func.(hd(list), acc)
    do_foldl(tl(list), acc, func, new_list ++ [value])
  end

  def flatten(list, tail \\ []) do
    do_flatten(list, []) ++ tail
  end

  defp do_flatten([], flattened_list) do
    flattened_list
  end

  defp do_flatten(list, flattened_list) do
    updated = case hd(list) do
      l when is_list(l) ->
        flattened_list ++ do_flatten(l, [])
      item ->
        flattened_list ++ [item]
    end

    do_flatten(tl(list), updated)
  end


  def keydelete(list, key, position) do
    do_keydelete(list, key, position, [])
  end

  defp do_keydelete([], _, _, new_list) do
    new_list
  end

  defp do_keydelete(list, key, position, new_list) do
    current_value = hd(list)

    updated = if elem(current_value, position) == key do
      new_list
    else
      new_list ++ [current_value]
    end

    do_keydelete(tl(list), key, position, updated)
  end

  def keyfind(list, key, position) do
    do_keyfind(list, key, position, nil)
  end

  def keyfind(list, key, position, default) do
    do_keyfind(list, key, position, default)
  end

  defp do_keyfind([], _, _, default) do
    default
  end

  defp do_keyfind(list, key, position, default) do
    current_value = hd(list)

    if elem(current_value, position) == key do
      current_value
    else
      do_keyfind(tl(list), key, position, default)
    end
  end

  def keymember?(list, key, position) do
    keyfind(list, key, position) != nil
  end

  def keyreplace(list, key, position, new_tuple) do
    do_keyreplace(list, key, position, [], new_tuple)
  end

  defp do_keyreplace([], _, _, new_list, _) do
    new_list
  end

  defp do_keyreplace(list, key, position, new_list, new_tuple) do
    current_value = hd(list)

    updated = if elem(current_value, position) == key do
      new_list ++ [new_tuple]
    else
      new_list ++ [current_value]
    end

    do_keyreplace(tl(list), key, position, updated, new_tuple)
  end

  def zip([]) do
    []
  end

  def zip(list_of_lists) when is_list(list_of_lists) do
    lengths = Enum.map(list_of_lists, fn(list) -> length(list) end)
    length = apply(JS.Math, :min, lengths)
    do_zip(list_of_lists, 0, length, [])
  end

  defp do_zip(list_of_lists, index, length, acc) when index == length do
    acc
  end

  defp do_zip(list_of_lists, index, length, acc) do
    values = Enum.map(list_of_lists, fn(list) -> Enum.at(list, index) end)
    item = JS.new(Bootstrap.Core.Tuple, values)
    
    do_zip(list_of_lists, index + 1, length, acc ++ [item])
  end
end
